File: LeetCode/213. House Robber II.cpp
// Author : @AdarshGaur

class Solution {
public:
    int f(vector<int>& a, int& n, int start, int end){
        vector<int> dp(n+1, 0);
        dp[start] = a[start-1];
        int ret = dp[start];
        for(int i=start+1; i<=end; i++){
            if(i < 3){
                dp[i] = a[i-1] + dp[i-2];
                continue;
            }
            dp[i] = max(dp[i-2], dp[i-3]) + a[i-1];
            ret = max(ret, dp[i]);
        }
        return ret;
    }
    int rob(vector<int>& nums) {
        int n = nums.size();
        if(n == 1) return nums.back();
        return max(f(nums, n, 1, n-1), f(nums, n, 2, n));
    }
};

File: LeetCode/213. House Robber II.java
class Solution {
    public int rob(int[] nums) {
        if(nums.length == 1)
            return nums[0];
        return Math.max(robCircle(nums, 1, nums.length-1), robCircle(nums, 0, nums.length-2));
    }

    public int robCircle(int[] nums, int start, int end) {
        int last = 0;
        int behindLast = 0;
        for(int i=start; i<=end; i++){
            int current = Math.max(behindLast + nums[i], last);
            behindLast = last;
            last = current;
        }
        return last;
    }
}

